File: code/ch19/19.4.3. 230.kthSmallest.java
class Solution {
  public int countNodes(TreeNode node) {
    if (node == null) return 0;
    int l = countNodes(node.left);
    int r = countNodes(node.right);
    return l + r + 1;
  }
  public int kthSmallest(TreeNode root, int k) {
    int cnt = countNodes(root.left);
    if (cnt == k - 1) return root.val;
    else if (cnt > k - 1)
      return kthSmallest(root.left, k);
    else
      return kthSmallest(root.right, k - cnt - 1);
  }
}

File: code/ch19/19.4.3. 230.kthSmallest.py
class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        # 返回该节点以及其所有子节点的个数
        def countNodes(node) -> int:
            if node == None:
                return 0
            l = countNodes(node.left)
            r = countNodes(node.right)
            return l + r + 1

        cnt = countNodes(root.left)
        if cnt == k - 1:
            return root.val
        elif cnt > k - 1:
            return self.kthSmallest(root.left, k)

        return self.kthSmallest(root.right, k - cnt - 1)
